File: Sorting and Searching/cpp/TrafficLights.cpp
#include <bits/stdc++.h>
using namespace std;

void calculateMaxDistances(int streetLength, int numberOfTrafficLights, int positionOfTrafficLights[]) {
    // set has default increasing ordering
    set<int> street{0, streetLength};
    // multiset
	multiset<int> passages{streetLength};
    
    for (int i = 0; i < numberOfTrafficLights; i++) {
        int trafficLightPosition = positionOfTrafficLights[i];
        auto nextLightPosition = street.upper_bound(trafficLightPosition);
        auto previousLightPosition = nextLightPosition;
        previousLightPosition--;

        passages.erase(passages.find(*nextLightPosition - *previousLightPosition));
		passages.insert(trafficLightPosition - *previousLightPosition);
		passages.insert(*nextLightPosition - trafficLightPosition);
		
        street.insert(trafficLightPosition);

		cout << *passages.rbegin() << " ";
    }

    return;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int streetLength = 0;
    cin >> streetLength;

    int numberOfTrafficLights = 0;
    cin >> numberOfTrafficLights;

    int positionOfTrafficLights[numberOfTrafficLights] = {0};
    for (int i = 0; i < numberOfTrafficLights; i++) {
        cin >> positionOfTrafficLights[i];
    }

    // solve
    calculateMaxDistances(streetLength, numberOfTrafficLights, positionOfTrafficLights);

    return 0;
}

File: Sorting and Searching/java/TrafficLights.java
import java.io.BufferedOutputStream;
import java.io.IOException;
import java.util.List;
import java.util.NavigableSet;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.TreeSet;

public class TrafficLights {
    
    public static void main(String[] args) throws IOException {

        // reader
        Scanner sc = new Scanner(System.in);

        // inputs
        int streetLength = sc.nextInt();
        int numberOfTrafficLights = sc.nextInt();
        sc.nextLine();

        String[] positionOfTrafficLights = sc.nextLine().split("\\s+");

        // writer
        BufferedOutputStream writer = new BufferedOutputStream(System.out);
        StringBuilder result = new StringBuilder();

        // solve
        NavigableSet<Integer> street = new TreeSet<>(List.of(0, streetLength));
        TreeMap<Integer, Integer> passages = new TreeMap<>();

        for (int i = 0; i < numberOfTrafficLights; i++) {
            Integer trafficLightPosition = Integer.parseInt(positionOfTrafficLights[i]);

            Integer nextLightPosition = street.higher(trafficLightPosition);
            Integer previousLightPosition = nextLightPosition != null ? street.lower(nextLightPosition) : null;

            if (previousLightPosition != null && nextLightPosition != null) {
                Integer passageToBeDivided = nextLightPosition - previousLightPosition;
                Integer leftDivision = trafficLightPosition - previousLightPosition;
                Integer rightDivision = nextLightPosition - trafficLightPosition;

                passages.put(passageToBeDivided, passages.getOrDefault(passageToBeDivided, 1) - 1);
                if (passages.getOrDefault(passageToBeDivided, 0) == 0) {
                    passages.remove(passageToBeDivided);
                }

                passages.put(leftDivision, passages.getOrDefault(leftDivision, 0) + 1);
                passages.put(rightDivision, passages.getOrDefault(rightDivision, 0) + 1);
            }

            street.add(trafficLightPosition);

            result.append(passages.lastKey()).append(" ");

        }

        writer.write(result.toString().getBytes());
        writer.flush();

        // close reader
        sc.close();
        
        return;

    }

}
